sympy/stats/rv_interface.py

```diff
@@ -339,6 +340,43 @@ def factorial_moment(X, n, condition=None, **kwargs):
     """
     return expectation(FallingFactorial(X, n), condition=condition, **kwargs)
 
+def median(X, evaluate=True, **kwargs):
+    """
+    Calculuates the median of the probability distribution.
+
+    Parameters
+    ==========
+
+    X: The random expression whose median is to be calculated.
+
+    Returns
+    =======
+
+    The median of the random expression.
+
+    Examples
+    ========
+
+    >>> from sympy.stats import Normal, Die, median
+    >>> N = Normal('N', 3, 1)
+    >>> median(N)
+    3
+    >>> D = Die('D')
+    >>> median(D)
+    3
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Median#Probability_distributions
+
+    """
+    result = quantile(X, evaluate=evaluate, **kwargs)(1/2)
+    if isinstance(result, FiniteSet):
+        for elem in result:
+                if cdf(X)(elem).round(8) == Float(1/2, 8):
```

Comment:
You are using float `1/2` here. It's best to use exact rationals like `Rational(1, 2)` or `S.Half`. We shouldn't need to use round here if exact results are used.

Reply:
This seems like an inefficient way to compute the median.

Reply:
@Smit-create Can we use `result` to do this computation?Are there any examples which produce `FiniteSet` on using `quantile(X, evaluate=evaluate, **kwargs)(1/2)`?

Reply:
Yes, There are certain distributions which gives `FiniteSet` and it basically does is `solve(cdf(X), 1/2, rv)` which may return finite set in case of square roots, so it is necessary to verify the exact roots.
```
>>> from sympy.stats import *
>>> quantile(Pareto('b',5,2))(1/2)
FiniteSet(-7.07106781186548, 7.07106781186548)
>>> quantile(Nakagami('n',1,1))(1/2)
FiniteSet(-0.832554611157698, 0.832554611157698)

```

Reply:
> This seems like an inefficient way to compute the median.

Actually, I tried with many ways of comparison, but Rational(1, 2) was completely out of my mind. I will correct the rounding off part with appropriate comparison

Reply:
I tried with:
```
result = quantile(X, evaluate=evaluate, **kwargs)(1/2)
    if isinstance(result, FiniteSet):
        for elem in result:
                if cdf(X)(elem) == Rational(1, 2):
                    return elem
    return result
```
The following are results:
```
>>> median(Nakagami('x',1,1))
FiniteSet(-0.832554611157698, 0.832554611157698)
>>> cdf(Nakagami('x',1,1))(0.832554611157698)
0.500000000000000
>>> cdf(Nakagami('x',1,1))(0.832554611157698)==Rational(1,2)
False
>>> float(cdf(Nakagami('x',1,1))(0.832554611157698))
0.5000000000000002
>>> float(Rational(1,2))
0.5

```

Reply:
@oscarbenjamin @asmeurer @czgdp1807  Can you please help me to get the correct comparison in the above reference?

Reply:
> There are certain distributions which gives `FiniteSet`

Is that even supposed to happen? It seems the FiniteSet is generated in the default/fallback routine for those distributions that don't define their own `_quantile` method:
https://github.com/sympy/sympy/blob/f8e7019feded4e4d0852f49fe0ccfa777f0e8744/sympy/stats/crv.py#L294-L317
According to codecov.io the whole `compute_quantile` method isn't tested at all.
I'd say `compute_quantile` should check if the solution set is a singleton FiniteSet (and if so return the element instead of the set) and raise an exception otherwise. (NotImplementedError?)

Reply:
You are still using a float `1/2`.

Reply:
Also: the call to solveset in line 305 above should use `self.set` instead of `S.Reals`. (That eliminates the spurious solution.)